File: dart/betterkhmer/test/fixtures_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:betterkhmer/betterkhmer.dart';

List<String> loadLines(String path) =>
    File(path).readAsLinesSync();

void main() {
  final fixturesDir = '${Directory.current.path}/../../fixtures';

  test('basic', () {
    expect(normalize('ខ្មែរ'), equals('ខ្មែរ'));
  });

  test('fixtures', () {
    final inputs   = loadLines('$fixturesDir/input.txt');
    final expected = loadLines('$fixturesDir/expected.txt');
    expect(inputs.length, equals(expected.length));

    var failures = 0;
    for (var i = 0; i < inputs.length; i++) {
      final got = normalize(inputs[i]);
      if (got != expected[i]) {
        failures++;
        if (failures <= 10) {
          print('[$i] got ${got.runes.map((r) => r.toRadixString(16)).join(' ')}, '
                'want ${expected[i].runes.map((r) => r.toRadixString(16)).join(' ')}');
        }
      }
    }
    expect(failures, equals(0), reason: '$failures/${inputs.length} fixture failures');
  });
}

File: rust/betterkhmer/tests/fixtures_test.rs
use std::fs;
use std::path::Path;

fn load_lines(path: &Path) -> Vec<String> {
    fs::read_to_string(path)
        .expect("cannot read file")
        .lines()
        .map(|l| l.to_string())
        .collect()
}

#[test]
fn test_basic() {
    assert_eq!(betterkhmer::normalize("ខ្មែរ"), "ខ្មែរ");
}

#[test]
fn test_fixtures() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent().unwrap()
        .parent().unwrap()
        .join("fixtures");

    let inputs   = load_lines(&root.join("input.txt"));
    let expected = load_lines(&root.join("expected.txt"));
    assert_eq!(inputs.len(), expected.len());

    let mut failures = 0usize;
    for (i, (inp, exp)) in inputs.iter().zip(expected.iter()).enumerate() {
        let got = betterkhmer::normalize(inp);
        if &got != exp {
            failures += 1;
            if failures <= 10 {
                eprintln!("[{i}] got {got:?}, want {exp:?} (input {inp:?})");
            }
        }
    }
    assert_eq!(failures, 0, "{failures}/{} fixture failures", inputs.len());
}
